### PR TITLE
planner: fix projection of generate column for indexLookup task

### LINKED ISSUE #31682: hash table + expression index, query return error msg “Internal error: UnionExec chunk column count mismatch, req: 2, result: 3”
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
drop table t1,t2;
create table t1 (c_int int, c_str varchar(40) ,primary key (c_int)  , key(c_str(36))  , key((c_int + 1))) partition by hash (c_int) partitions 4 ;
create table t2  like t1 ;
insert into t1 values (1, 'sleepy kowalevski');
insert into t2 values (3, 'unruffled chaplygin');
select (select t2.c_str from t2 where t2.c_int + 1 = 4 order by t2.c_str) x from t1; 
select (select t2.c_str from t2 where t2.c_int = 3 order by t2.c_str) x from t1; 
```

### 2. What did you expect to see? (Required)

```
mysql> select (select t2.c_str from t2 where t2.c_int + 1 = 4 order by t2.c_str) x from t1;
+---------------------+
| x                   |
+---------------------+
| unruffled chaplygin |
+---------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)

```
mysql> select (select t2.c_str from t2 where t2.c_int + 1 = 4 order by t2.c_str) x from t1;
ERROR 1105 (HY000): Internal error: UnionExec chunk column count mismatch, req: 2, result: 3

mysql> select (select t2.c_str from t2 where t2.c_int = 3 order by t2.c_str) x from t1;     
+---------------------+
| x                   |
+---------------------+
| unruffled chaplygin |
+---------------------+
1 row in set (0.01 sec)

```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.4.0-nightly
Edition: Community
Git Commit Hash: 4baab3cf1abdf09cf7b5dc6078b770cf67905aff
Git Branch: heads/refs/tags/v5.4.0-nightly
UTC Build Time: 2022-01-10 05:19:54
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #31682

Problem Summary:
We don't handle the needExtraProj correctly. If the task is an IndexLookup task, it would miss the extra projection.

### What is changed and how it works?
use t.needExtraProj and t.originalSchema

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by time-and-fate — APPROVED** [review]

File: planner/core/task.go [hunk 1]
```diff
@@ -1035,11 +1033,15 @@ func (t *copTask) convertToRootTaskImpl(ctx sessionctx.Context) *rootTask {
...
-		prevSchema = ts.schema.Clone()
+		prevSchema := ts.schema.Clone()
 		ts.Columns = ExpandVirtualColumn(ts.Columns, ts.schema, ts.Table.Columns)
 		if len(ts.Columns) > prevColumnLen {
 			// Add an projection to make sure not to output extract columns.
-			needExtraProj = true
+			t.needExtraProj = true
+			// Keep originSchema if exists so that we don't overlap it.
+			if t.originSchema == nil {
+				t.originSchema = prevSchema
```
Comment (mjonss) [inline planner/core/task.go:1043]: Why not remove `prevSchema` variable and replace it here with `ts.schema.Clone()`?

**Review by mjonss — COMMENTED** [review]
Just a small suggestion.

File: planner/core/task.go (hunk 1 above)
Comment (wjhuang2016) [reply planner/core/task.go:1043]: ExpandVirtualColumn would change the schema.

Comment (mjonss) [reply planner/core/task.go:1043]: Ah, thanks for the explanation!

**Review by mjonss — APPROVED** [review]
LGTM

File: planner/core/task.go [hunk 2]
```diff
@@ -1035,11 +1033,15 @@ func (t *copTask) convertToRootTaskImpl(ctx sessionctx.Context) *rootTask {
 		}
 		ts := tp.(*PhysicalTableScan)
 		prevColumnLen := len(ts.Columns)
-		prevSchema = ts.schema.Clone()
+		prevSchema := ts.schema.Clone()
 		ts.Columns = ExpandVirtualColumn(ts.Columns, ts.schema, ts.Table.Columns)
 		if len(ts.Columns) > prevColumnLen {
```
Comment (winoros) [inline planner/core/task.go:1038]: Can we directly change the condition to `t.needExtraProj == false || len(ts.Columns) > prevColumnLen`?

**Review by winoros — APPROVED** [review]

**Comment by time-and-fate** [comment]
This PR triggers https://github.com/pingcap/tidb/issues/33237, which is expected to be fixed by https://github.com/pingcap/tidb/pull/33287.
Please take care when cherry-picking.

**Pull request merged**